### ISSUE TITLE
RFE: Add support for soft/virtual deletes at server level.

### ISSUE THREAD

**Issue by zakaluka** [Kinto/kinto#1352]
There are certain use cases where it is not permissible to "hard delete" anything from the database.  Instead, everything must be "virtual deleted" only.  This functionality would be similar to Kinto.js' virtual delete support (but not exactly because Kinto.js will "hard delete" a virtually deleted item on sync).

This request is to support the use cases mentioned above at the server level in one of the following ways:

1. Separate 'write' permissions into 'update' and 'delete'.  This would break backwards compatibility (unless 'write' were kept on in addition to 'update' and 'delete').

1. Allow an administrator to enable/disable HTTP verbs (e.g. DELETE in this instance).  I don't even know if this is possible, as I'm just starting to learn / use Kinto.

    An alternative to this strategy would be to block verbs at a firewall level, though I would prefer this solution to be within Kinto rather than relying on other pieces of software.

1. Provide "virtual delete" as an option that would change the server's default behavior on delete to mark an item as 'deleted' and, maybe, hide such items by default in queries.  This would require adding an attribute to every item (a status like Kinto.js' `_status` with values such as "active" and "deleted").  This would keep backwards compatibility if the server were to assume that any item that is missing this attribute is in Active status.

    I don't know much about Kinto plugins.  However, if this option can be implemented as a plugin, I would be willing to try and take a crack at it.  However, not really knowing Kinto or Python means that hopefully there's a really good tutorial (or sample plugin) out there that teaches you how to do such things.

The purpose of this post is to hopefully start a discussion on this topic and, eventually, provide such functionality within the Kinto server.

**Comment by Natim** [comment]
We already do that for encrypted content were we set the status to deleted in an encrypted manner and then kinto.js catches it or in the blocklist with an enabled/disabled field.

You can change the metadata to do the same, you can also use the history plugin to retrieve a previous version of a deleted record.

What is your use case for virtually deleted?

**Comment by zakaluka** [comment]
The use case is a fully auditable system where every user's every action (including views) are being logged.  A supervisor/administrator must be able to come in at any time and view all records, including deleted ones (with full history such as values changed, dates of change, etc.).  A `GET` on a bucket / collection / record, rather than filtering / aggregating history results, seemed to be a much easier option from a end-user perspective.

I read about the history plugin (just on the website) but wasn't sure whether it would be sufficient for this purpose.

Two questions:
1. I'm not sure what you mean by deleting the encrypted content in an encrypted manner.  Isn't encryption a kinto.js-only concept?
    Note: I'm not familiar enough with the Kinto-HTTP interface to know if it supports encryption or not.  I'm making an assumption that any web-service type interface relies on SSL for protection (rather than encrypting the data as well).

1. I assume you are referring to the bucket/collection/record metadata.  If so, I am leaning towards setting a status as deleted in that metadata (rather than actually running a DELETE command against the database).  However, my concern is about what would happen if a DELETE command ended up being sent to the server.  How do I stop it from performing its default action?  That was my main reason for suggesting that 'write' permissions should be split into 'update' and 'delete' as 2 separate permissions.

**Comment by Natim** [comment]
Thank you for your responses, really pleasant to read.

> The use case is a fully auditable system where every user's every action (including views) are being logged. A supervisor/administrator must be able to come in at any time and view all records, including deleted ones (with full history such as values changed, dates of change, etc.).

If your aim is to allow review of actions, then what you want is exactly the history plugin. It is also well integrated with the kinto admin, it allows to see who did what action as well as showing a diff between two dates of what happened.

> what you mean by deleting the encrypted content in an encrypted manner. Isn't encryption a kinto.js-only concept?

Correct, encryption always happens on the client side.
You use it when you don't want the administrator of the database to be able to read the content stored for privacy reasons.

In that setup, you might want to encrypt the fact that a record was deleted to prevent the administrator to know how many records were used and how many were deleted because this information could be sufficient to find out what the user is storing in the collection.

When we do that, we don't use the DELETE verb but we encrypt the `{status: deleted}` metadata. Kinto.js detects it as a deleted item but in that case we actually don't delete the record.

**Comment by Natim** [comment]
> I also didn't fully understand what blocklists are

The blocklist is yet another use-case we use Kinto for.

For instance: https://firefox.settings.services.mozilla.com/v1/buckets/blocklists/collections/certificates/records

You can see that records contains a `enabled` field that allow us to deactivate them without removing them.

**Comment by Natim** [comment]
You can deactivate some endpoint verbs using: http://kinto.readthedocs.io/en/stable/configuration/settings.html#enabling-or-disabling-endpoints

```

kinto.collection_bucket_delete_enabled = false
kinto.record_bucket_delete_enabled = false

kinto.collection_collection_delete_enabled = false
kinto.record_collection_delete_enabled = false

kinto.collection_record_delete_enabled = false
kinto.record_record_delete_enabled = false
```

**Comment by zakaluka** [comment]
@Natim Thank you for all the detailed answers.  They were extremely informative and also helped show how much more I need to learn about Kinto :).  

> You can deactivate some endpoint verbs using: http://kinto.readthedocs.io/en/stable/configuration/settings.html#enabling-or-disabling-endpoints

I can't believe I missed that section! It is doing precisely what I needed to do (or at least part of it).

> For instance: https://firefox.settings.services.mozilla.com/v1/buckets/blocklists/collections/certificates/records
> 
> You can see that records contains a enabled field that allow us to deactivate them without removing them.

This is also an excellent option for what I am interested in doing.  It is also highly preferable because it allows me to have full control over how to treat a record that is disabled / "deleted".  

From the information you provided, a combination of disabling the delete endpoints and using a technique similar to Mozilla's use with blocklists will give me exactly what I set out to find.

Thank you again for all your help.  I will close this issue now but I am wondering whether it is worthwhile to add a section to the manual that contains the "techniques" you explained with a description / examples of what they do (for someone else who may come looking).
